src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs

```diff
@@ -286,66 +286,62 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+                // We've managed to reach the use location
+                debug!("can_reach_ignoring_backedges: found!");
                 return true;
             }
 
             // Skip locations we've been.
-            if visited_locations.contains(&location) { continue; }
+            if visited_locations.contains(&location) {
+                continue;
+            }
 
             let block = &self.mir.basic_blocks()[location.block];
-            if location.statement_index ==  block.statements.len() {
-                // Add start location of the next blocks to pending locations.
-                match block.terminator().kind {
-                    TerminatorKind::Goto { target } => {
-                        pending_locations.push(target.start_location());
-                    },
-                    TerminatorKind::SwitchInt { ref targets, .. } => {
-                        pending_locations.extend(
-                            targets.into_iter().map(|target| target.start_location()));
-                    },
-                    TerminatorKind::Drop { target, unwind, .. } |
-                    TerminatorKind::DropAndReplace { target, unwind, .. } |
-                    TerminatorKind::Assert { target, cleanup: unwind, .. } |
-                    TerminatorKind::Yield { resume: target, drop: unwind, .. } |
-                    TerminatorKind::FalseUnwind { real_target: target, unwind, .. } => {
-                        pending_locations.push(target.start_location());
-                        if let Some(unwind) = unwind {
-                            pending_locations.push(unwind.start_location());
-                        }
-                    },
-                    TerminatorKind::Call { ref destination, cleanup, .. } => {
-                        if let Some((_, destination)) = destination {
-                            pending_locations.push(destination.start_location());
-                        }
-                        if let Some(cleanup) = cleanup {
-                            pending_locations.push(cleanup.start_location());
-                        }
-                    },
-                    TerminatorKind::FalseEdges { real_target, ref imaginary_targets, .. } => {
-                        pending_locations.push(real_target.start_location());
-                        pending_locations.extend(
-                            imaginary_targets.into_iter().map(|target| target.start_location()));
-                    },
-                    _ => {},
-                }
-            } else {
+
+            if location.statement_index < block.statements.len() {
                 // Add the next statement to pending locations.
-                pending_locations.push(location.successor_within_block());
+                pending_locations.push((location.successor_within_block(), false));
+            } else {
+                pending_locations.extend(
```

Comment:
here you can just only add a successor to `pending_locations` if they are (a) not yet visited and (b) (p -> successor) is not a backedge

```rust
pending_locations.extend(
    block
        .terminator()
        .successors()
        .map(|bb| Location {
                                statement_index: 0,
                                block: bb,
                            })
        .filter(|s| visited_locations.insert(s)) // only things we haven't seen before
        .filter(|s| !location.dominates(s, &self.mir.dominators())) // not backedges
);
```